### PR TITLE
Sylvia whittle/431 ignore high grains

### LINKED ISSUE #431: Remove grains based on height
**Is your feature request related to a problem? Please describe.**
It would be useful to be able to remove grains for being too high. This could prevent certain impurities having the same area as molecules of interest from being selected.

**Describe the solution you'd like**
Introduce another user configurable threshold based on the mean height of the grain, so that for example, grains whose mean height deviates from the image mean by over 5 sigma can be excluded. (Suggested by @SylviaWhittle ) 
Example images will be supplied as google drive links in the comments.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by SylviaWhittle**
Closes #431 

Adds a step to `grains.py` that removes grains from the mask based on their median heights. This enables users to specifically remove tall grains such as proteins when imaging DNA.

The thresholds for upper / lower limits for this removal are configurable, to enable as much configuration as possible. 

If not set, this will default to not removing any grains.

### CODE REVIEW

File: topostats/default_config.yaml [hunk 1]
```diff
@@ -54,9 +58,9 @@ plotting:
   run: true # Options : true, false
   save_format: png # Options : see https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.savefig.html
   pixel_interpolation: null # Options : https://matplotlib.org/stable/gallery/images_contours_and_fields/interpolation_methods.html
-  image_set: core  # Options : all, core
-  zrange: [null, null]  # low and high height range for core images (can take [null, null]). low <= high
-  colorbar: true  # Options : true, false
+  image_set: all # Options : all, core
```
Comment (ns-rse) [inline topostats/default_config.yaml:61]: Will go through in more detail over the coming days but I think the desired default here should be `core` as this file serves for output from `--create-config-file` and I thought most of the time people don't require all grain images.

File: topostats/validation.py [hunk 2]
```diff
@@ -200,6 +200,28 @@ def validate_config(config: dict, schema: Schema, config_type: str) -> None:
                 "above",
                 error="Invalid direction for grains.direction valid values are 'both', 'below' or 'above",
             ),
+            "grain_height_removal_thresholds_std_dev": {
+                "above": [
+                    Or(
+                        float,
```
Comment (ns-rse) [inline topostats/validation.py:206]: Should the `float` be within a certain range? E.g. always `> 0.0`? I'm unsure of the expected ranges but see that the tests have arrays with negative values so `float` may be perfectly valid here but wanted to check as I thought all values are typically `> 0.0` (although perhaps some materials return negative values, I really don't know/remember).

File: topostats/default_config.yaml (hunk 1 above)
Comment (SylviaWhittle) [reply topostats/default_config.yaml]: This change was pushed in error, I was testing output images and forgot to change this back, thanks!

File: topostats/validation.py (hunk 2 above)
Comment (SylviaWhittle) [reply topostats/validation.py]: I think it would be sensible to set it so that it can only be positive. I don't see any reason why a negative value would be valid, as if you set the thresholds for above to be negative, that means that you're looking for grains detected above the image average, which are below the image average. It's nonsensical 👍. I have added lambda statements.

**Review by Jean-Du — CHANGES_REQUESTED** [review]
Hi, thanks a lot for adding this feature! I ran it through a test image and set the maximum threshold of 'above' to 100, thinking that it should only remove grains whose mean height is higher than 100 standard deviations away from the median; for some reason, this resulted in all the grains being removed. Did I misunderstand what the parameters meant?
I also had a quick look at the changes and made some minor comments.

File: topostats/default_config.yaml [hunk 3]
```diff
@@ -42,8 +42,12 @@ grains:
   # Thresholding by area
   smallest_grain_size_nm2: 50 # Size in nm^2 of tiny grains/blobs (noise) to remove, must be > 0.0
   absolute_area_threshold:
-    above: [ 300, 3000 ] # above surface [Low, High] in nm^2 (also takes null)
-    below: [ null, null ] # below surface [Low, High] in nm^2 (also takes null)
+    above: [0, 10000000] # above surface [Low, High] in nm^2 (also takes null)
+    below: [null, null] # below surface [Low, High] in nm^2 (also takes null)
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/default_config.yaml:46]: The `300` to `3000` range was only really suiting `minicircle.spm` style images, and so in an effort to make the software more generally applicable by default, I've set it so that it won't remove any grains.

[the remark this answers is not in the document]
Comment (Jean-Du) [reply topostats/default_config.yaml:50]: In that case, would it make more sense to just set it to null?

File: topostats/default_config.yaml [hunk 4]
```diff
@@ -42,8 +42,12 @@ grains:
...
+  # Standard deviation multiplier for height thresholds for grain median heights. Removes grains whose median is [higher, lower] than these values.
+  grain_height_removal_thresholds_std_dev:
+    below: [null, null] # Minimum, maximum std dev multiplier for grain median heights for the below direction.
+    above: [null, null] # Minimum, maximum std dev multiplier for grain median heights for the above direction.
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/default_config.yaml]: Done 👍

File: topostats/default_config.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/default_config.yaml]: Good point, done 👍

File: topostats/plotting_dictionary.yaml [hunk 5]
```diff
@@ -109,27 +109,32 @@ removed_small_objects:
   title: "Small Objects Removed"
   image_type: "binary"
   core_set: false
+grains_removed_based_on_height:
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/plotting_dictionary.yaml:112]: I think that `remove_small_objects` may be best deleted, since so much of what it does overlaps with `area_thresholding`, so I think I'll leave this to discussion and a different PR if that's okay

**Comment by SylviaWhittle** [comment]
Thanks for your comments @Jean-Du , I've fixed the bug that caused the problem you found. It was a single minus sign.

File: topostats/default_config.yaml [hunk 6]
```diff
@@ -39,11 +39,15 @@ grains:
...
   # Thresholding by area
   smallest_grain_size_nm2: 50 # Size in nm^2 of tiny grains/blobs (noise) to remove, must be > 0.0
   absolute_area_threshold:
-    above: [ 300, 3000 ] # above surface [Low, High] in nm^2 (also takes null)
-    below: [ null, null ] # below surface [Low, High] in nm^2 (also takes null)
+    above: [null, null] # above surface [Low, High] in nm^2 (also takes null)
+    below: [null, null] # below surface [Low, High] in nm^2 (also takes null)
```
Comment (Jean-Du) [inline topostats/default_config.yaml:50]: Just something minor, but I noticed that all the other sections have 'below' first and 'above' second, and here their order is inversed which could cause confusion

**Review by Jean-Du — CHANGES_REQUESTED** [review]
Sorry this is so late! I ran it through again on my test file and for some reason it still seems to be behaving strangely:
My test file has a grain that's much higher than others, and I am hoping that this particular grain can be excluded while other grains are kept; but when I changed the upper threshold of the 'above' grains, it seems that either everything is masked, or nothing is masked. The cutoff is somewhere between 2.7 and 2.8. I can message you the file I use for testing in private if it helps!

**Comment by Jean-Du** [comment]
Config file for reference:
```
# Configuration from TopoStats run completed : 2023-06-01 17:44:31
# For more information on configuration : https://afm-spm.github.io/TopoStats/main/configuration.html
base_dir: G:\My Drive\PhD\Data\HU_project\Tests\High_objects_test
output_dir: output_new
log_level: info
cores: 2
file_ext: .spm
loading:
  channel: Height
filter:
  run: true
  row_alignment_quantile: 0.5
  threshold_method: std_dev
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    below: 10.0
    above: 1.0
  threshold_absolute:
    below: -1.0
    above: 1.0
  gaussian_size: 1.0121397464510862
  gaussian_mode: nearest
  remove_scars:
    run: true
    removal_iterations: 2
    threshold_low: 0.25
    threshold_high: 0.666
    max_scar_width: 4
    min_scar_length: 16
grains:
  run: true
  threshold_method: std_dev
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    below: 10.0
    above: 1.0
  threshold_absolute:
    below: -1.0
    above: 1.0
  direction: above
  grain_height_removal_thresholds_std_dev:
    below:
    - 
    - 
    above:
    - 
    - 2.75
  smallest_grain_size_nm2: 50
  absolute_area_threshold:
    above:
    - 
    - 
    below:
    - 
    - 
grainstats:
  run: true
  edge_detection_method: binary_erosion
  cropped_size: 40.0
dnatracing:
  run: true
  min_skeleton_size: 10
plotting:
  run: true
  save_format: png
  pixel_interpolation:
  image_set: core
  zrange:
  - 
  - 
  colorbar: true
  axes: true
  cmap: nanoscope
  mask_cmap: blu
  histogram_log_axis: false
  histogram_bins: 200
  dpi: 100
summary_stats:
  run: true
  config:
```

**Comment by ns-rse** [comment]
Looking to close up old issues that are lingering and I'm curious if this might be addressed by #666 ?

**Comment by SylviaWhittle** [comment]
> Looking to close up old issues that are lingering and I'm curious if this might be addressed by #666 ?

This feature is slightly different. It removes grains whose median value is higher than a threshold. #666 prevents pixels above a threshold from being added to the mask.

In the former case, a protein that is very high up but has low pixels around the edge would be entirely removed. In the latter case, the ring of low pixels around the protein would be masked and kept as a grain. Does that make sense?

**Comment by ns-rse** [comment]
@SylviaWhittle will finish this off in the coming weeks as she is familiar with the work and it doesn't require too much more effort.

**Comment by SylviaWhittle** [comment]
Concept approved by experimentalists. Use max value. Users will be able to guess a good height to set as the maximum for their sample.

Revive this PR
